contributors/design-proposals/pod-resource-management.md

```diff
@@ -1,114 +1,367 @@
-  - [Rollout Plan](#rollout-plan)
-  - [Implementation Status](#implementation-status)
-
-<!-- END MUNGE: GENERATED_TOC -->
+# Kubelet pod level resource management
+
+**Authors**:
+
+1. Buddha Prakash (@dubstack)
+1. Vishnu Kannan (@vishh)
+1. Derek Carr (@derekwaynecarr)
+
+**Last Updated**: 01/30/2017
+
+**Status**: Implementation planned for Kubernetes 1.6
+
+This document proposes a design for introducing pod level resource accounting
+to Kubernetes. It outlines the implementation and associated rollout plan.
 
 ## Introduction
 
-As of now [Quality of Service(QoS)](../../docs/design/resource-qos.md) is not enforced at a pod level. Excepting pod evictions, all the other QoS features are not applicable at the pod level.
-To better support QoS, there is a need to add support for pod level resource accounting in Kubernetes.
+Kubernetes supports container level isolation by allowing users
+to specify [compute resource requirements](resources.md) via requests and
+limits on individual containers.  The `kubelet` delegates creation of a
+cgroup sandbox for each container to its associated container runtime.
+
+Each pod has an associated [Quality of Service (QoS)](resource-qos.md)
+class based on the aggregate resource requirements made by individual
+containers in the pod.  The `kubelet` has the ability to
+[evict pods](kubelet-eviction.md) when compute resources are scarce. It evicts
+pods with the lowest QoS class in order to attempt to maintain stability of the
+node.
+
+The `kubelet` has no associated cgroup sandbox for individual QoS classes or
+individual pods.  This inhibits the ability to perform proper resource
+accounting on the node, and introduces a number of code complexities when
+trying to build features around QoS.
+
+This design introduces a unified cgroup hierarchy to enable the following:
+
+1. Improve enforcement of QoS class on the node.
+1. Simplify resource accounting at the pod level.
+1. Allow containers in a pod to share slack resources.  A pod with a Burstable
+container and a BestEffort container is classified as a Burstable pod.  The
+BestEffort container is not able to consume slack resources from the sibling
+Burstable container.  It must instead compete for scarce resources at the node
+level across all containers in all QoS classes which violates the design.
+1. Ability to charge per container overhead to the pod instead of the node.
+This overhead is container runtime specific.  For example, `docker` has
+an associated `containerd-shim` process that is created for each container
+which should be charged to the pod.
+1. Ability to charge any memory usage of memory-backed volumes to the pod when
+an individual container exits instead of the node.
+
+## Enabling the unified cgroup hierarchy
+
+To enable the unified cgroup hierarchy, the operator must enable the
+`--cgroups-per-qos` flag.  Once enabled, the `kubelet` will start managing
```

Comment:
per node allocatable design, i am keeping the cgroup-root flag defaulted to / since we will create the kubepods cgroup relative to that.